File: Heap/Python/Meeting Rooms II.py
# Problem link: https://leetcode.com/problems/meeting-rooms-ii/description/
# Time: O(n log n)
# Space: O(n)
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        intervals.sort()
        minHeap = []

        for start, end in intervals:
            if minHeap and minHeap[0] < start:
                heapq.heappop(minHeap)
            
            heapq.heappush(minHeap, end)

        return len(minHeap)
            

File: Heap/Java/Meeting Rooms II.java
// Problem link: https://leetcode.com/problems/meeting-rooms-ii/description/
// Time: O(n log n)
// Space: O(n)
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();

        for(int[] pair : intervals) {
            if(!minHeap.isEmpty() && minHeap.peek() < pair[0]) {
                minHeap.poll();
            }

            minHeap.add(pair[1]);
        }

        return minHeap.size();
    }
}
